localboot/main.go

```diff
@@ -110,13 +87,116 @@ func main() {
 		}
 	}
 	log.Print("No boot configuration succeeded")
-
-	// if we are here, booting failed, so let's clean things up by closing all
-	// open file descriptors and unmounting the devices that we mounted
-	for _, cfg := range bootconfigs {
-		cfg.Close()
-	}
+	// clean up
 	for _, mountpoint := range mounted {
 		syscall.Unmount(mountpoint.Path, syscall.MNT_DETACH)
 	}
+	return nil
+}
+
+// BootPathMode tries to boot a kernel in PATH mode. This means:
+// * look for a partition with the given GUID and mount it
+// * look for the kernel and initramfs in the provided locations
+// * boot the kernel with the provided command line
+//
+// The first parameter, `devices`, is a list of `storage.BlockDev` objects to
+// look for the given GUID. The second parameter, `guid`, is the partition GUID
+// to look for. The third, `dryrun`, will not actually boot the found
+// configuration if set to true.
+func BootPathMode(devices []storage.BlockDev, baseMountpoint string, guid string, dryrun bool) error {
+	debug("Getting list of supported filesystems")
+	filesystems, err := storage.GetSupportedFilesystems()
+	if err != nil {
+		log.Fatal(err)
+	}
+	debug("Supported file systems: %v", filesystems)
+
+	log.Printf("Looking for partition with GUID %s", guid)
+	partitions, err := storage.PartitionsByGUID(devices, guid)
+	if err != nil || len(partitions) == 0 {
+		return fmt.Errorf("Error looking up for partition with GUID %s", guid)
+	}
+	log.Printf("Partitions with GUID %s: %+v", guid, partitions)
+	if len(partitions) > 1 {
+		log.Printf("Warning: more than one partition found with the given GUID. Using the first one")
+	}
+	dev := partitions[0]
+	mountpath := path.Join(baseMountpoint, dev.Name)
+	devname := path.Join("/dev", dev.Name)
+	if _, err := storage.Mount(devname, mountpath, filesystems); err != nil {
+		return fmt.Errorf("Cannot mount %s on %s: %v", dev.Name, mountpath, err)
+	}
+	fullKernelPath := path.Join(mountpath, *kernelPath)
+	fullInitramfsPath := path.Join(mountpath, *initramfsPath)
+	kernelFD, err := os.Open(fullKernelPath)
+	if err != nil {
+		return fmt.Errorf("Cannot open kernel %s: %v", fullKernelPath, err)
+	}
+	initramfsFD, err := os.Open(fullInitramfsPath)
+	if err != nil {
+		return fmt.Errorf("Cannot open initramfs %s: %v", fullInitramfsPath, err)
+	}
+	cfg := BootConfig{
+		Kernel:    kernelFD,
+		Initramfs: initramfsFD,
+		Cmdline:   *kernelCmdline,
+	}
+	debug("Trying boot configuration %+v", cfg)
+	if dryrun {
+		log.Printf("Dry-run, will not actually boot")
+	} else {
+		if err := cfg.Boot(); err != nil {
+			cfg.Close()
+			return fmt.Errorf("Failed to boot kernel %s: %v", cfg.Kernel.Name(), err)
```

Comment:
Is there a chance we ever get here? Do you want to close the files from above before returning in this case? (kernelFD and initramfsFD)

Reply:
yes, the boot config may fail and I want to close file descriptors that are not going to be used anymore